Input: Number type input

alexis-ld opened the pull request:
**Component**:
Input

<!-- E.g. 'Layout', 'Table', 'build', 'tests'... -->

**Description**:

This adds a number type input as specified in #211  .

**Design**:

I kept the existing Input design for consistency and made it work with type="number" cases.

![Screenshot 2020-12-15 at 15 56 14](https://user-images.githubusercontent.com/3480526/102231539-5f7b8080-3eee-11eb-8be8-dab82ffe9642.png)




---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #211 

<!-- If you want to refer to an issue while not closing it, use:


-->

bert-e commented:
# Hello alexis-ld,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/core-ui/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @nicolas2bert

* @rachedbenmustapha

* @monpote

* @chengyanjin

ChengYanJin commented on stories/input.js:
```diff
@@ -95,18 +96,27 @@ const ExampleInput = () => {
         cols={50}
         rows={5}
       />
+      <Title>Number Inputl</Title>
```
```suggestion
      <Title>Number Input</Title>
```

ChengYanJin commented on src/lib/components/input/Input.component.js:
```diff
@@ -1,19 +1,19 @@
-import Checkbox from "../checkbox/Checkbox.component";
-import Select from "../select/Select.component";
-import TextArea from "../textarea/TextArea.component";
+import React, { Fragment } from 'react';
```
FYI, To simplify the code, we could use short syntax. See below,
https://reactjs.org/docs/fragments.html#short-syntax

But looks like you did not use Fragment in the code, should be removed.

ChengYanJin commented on src/lib/components/input/Input.component.js:
```diff
@@ -1,19 +1,19 @@
+import { DebounceInput } from 'react-debounce-input';
+import Checkbox from '../checkbox/Checkbox.component';
+import Select from '../select/Select.component';
+import TextArea from '../textarea/TextArea.component';
```
Ha!! I realized that in core-ui, we use double quotes. 
I would suggest to remove   `"singleQuote": true,` from .prettierrc. Otherwise, it will bring too much noisy.

ChengYanJin commented on src/lib/components/input/Input.component.style.js:
```diff
@@ -28,12 +28,35 @@ export const InputContainer = styled.div`
+    &::-webkit-inner-spin-button {
+      opacity: 0; // Disable default number input arrows
+    }
+  }
```
We should take care of Firefox as well.

ChengYanJin commented on src/lib/components/input/Input.component.style.js:
```diff
@@ -28,12 +28,35 @@ export const InputContainer = styled.div`
+      transform: translate(-50%, -50%);
+    }
+  }
+
```
I understand you tried to set the opacity:0 to the original arrows so that we can't see them, but the functionality is still there. 

But it does not work well on my side. I need to be at a little left to the new icon `<i className="fas fa-sort"></i>` I am wondering is it working well on your side?

ChengYanJin commented on stories/input.js:
```diff
@@ -95,18 +96,27 @@ const ExampleInput = () => {
+      <Input
+        id="id8"
+        value={numberValue}
+        onChange={(e) => setNumberValue(e.target.value)}
```
This cause a flow error in CI.
Something need to be change in the props definition.

  ```js
  onChange: (any) => void,
  ```
  
> Cannot get `e.target` because property `target` is missing in undefined [1].
>    stories/input.js:111:41
>    111|         onChange={(e) => setNumberValue(e.target.value)}

ChengYanJin reviewed (changes requested):
Some changes required. 

Most importantly, please test the arrows to see if increment and decrement work.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @nicolas2bert

* @rachedbenmustapha

* @monpote

* @chengyanjin





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

bert-e commented:
# Conflict




There is a conflict between your branch `feature/number-input` and the
destination branch `development/1.0`.

Please resolve the conflict on **the feature branch** (`feature/number-input`).

```bash
 $ git fetch
 $ git checkout origin/feature/number-input
 $ git merge origin/development/1.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:feature/number-input
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @nicolas2bert

* @rachedbenmustapha

* @monpote

* @chengyanjin





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

ChengYanJin commented on src/lib/components/input/Input.component.js:
```diff
@@ -37,8 +39,44 @@ const InputRenderer = ({
+        if (e.target.dataset.core === "up")
+          nativeInputValueSetter.call(
+            inputEl.current,
+            max && parseInt(value) + 1 <= parseInt(max)
```
if I don't specify `max`, I may not be able to increment?

alexis-ld replied:
Indeed... fixing

alexis-ld committed:
Prettier: remove singleQuote statement

alexis-ld committed:
Input: Add number input type

Refs: #211

alexis-ld committed:
Input: Add number input to existing Input story

alexis-ld committed:
Input: Add tests snapshots for number input

alexis-ld commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @nicolas2bert

* @rachedbenmustapha

* @monpote

* @chengyanjin





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin







*The following options are set:* **approve**

ChengYanJin reviewed (changes requested):
Please add an error state number input in the storybook, and fix the bug of the carets are outside the input component. 
 
 Thanks

alexis-ld committed:
Input: Update wrapper positionning

Moved relative positioning to number input wrapper instead of global input wrapper. The initial positionning was causing caret displacement when an error was specified.

Refs: #211

alexis-ld committed:
Input: Update test after input story change

ChengYanJin committed:
input: Improve the input number story with error

Refs: #211

alexis-ld committed:
Input: Change input wrapper align to baseline when input is a number

alexis-ld committed:
Input: update test after flexbox CSS change

ChengYanJin reviewed (approved):
LGTM!
Great solution `align-items: baseline;`!

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.0`




There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.0`




Please check the status of the associated issue None.

Goodbye alexis-ld.
